### ISSUE TITLE
"Found unexpected equivalence"

### ISSUE THREAD

**Issue by dylanede** [egraphs-good/egg#70]
I'm writing a very custom `Language` implementation and I suspect that maybe there is a bug in my implementation that is causing this assertion in "egraph.rs" to fail. Question is, I'm not sure what sort of problem leads to "Found unexpected equivalence", so maybe the text of that assertion could be expanded to include details of what the likely cause is?

**Comment by mwillsey** [comment]
Yeah, that's a little confounding. That check is to make sure that the rebuild process went correctly, I'm not sure how it could have failed. Do you have some funky eq or hash implementations for your Language?

**Comment by dylanede** [comment]
I do. I've found what the problem was. I was assuming that if a `Language` instance was created from the exact same syntax tree instance as another `Language` instance, then those two `Language` instances would always be equal - i.e. I had added an "optimisation" that allows me to short-circuit equality tests by skipping things like comparing the children `Id`s. I see now that since the children `Id`s are modified by egg, and that it relies on modified nodes no longer being equal to their past selves, this "optimisation" cannot work. I've removed it and the assertion failure is gone.

I think it might be worth mentioning something about the expected behaviour of equality on `Language`, though I appreciate I was taking a fairly liberal view of it.
